common/lib/dependabot/dependency.rb

```diff
@@ -119,61 +199,67 @@ def humanized_previous_version
         return ref_changed? ? previous_ref : nil
       end
 
-      if previous_version.match?(/^[0-9a-f]{40}/)
+      if T.must(previous_version).match?(/^[0-9a-f]{40}/)
         return previous_ref if ref_changed? && previous_ref
 
-        "`#{previous_version[0..6]}`"
+        "`#{T.must(previous_version)[0..6]}`"
       elsif version == previous_version &&
             package_manager == "docker"
-        digest = docker_digest_from_reqs(previous_requirements)
-        "`#{digest.split(':').last[0..6]}`"
+        digest = docker_digest_from_reqs(T.must(previous_requirements))
+        "`#{T.must(digest.split(':').last)[0..6]}`"
       else
         previous_version
       end
     end
 
+    sig { returns(T.nilable(String)) }
     def humanized_version
       return if removed?
 
-      if version.match?(/^[0-9a-f]{40}/)
+      if T.must(version).match?(/^[0-9a-f]{40}/)
         return new_ref if ref_changed? && new_ref
 
-        "`#{version[0..6]}`"
+        "`#{T.must(version)[0..6]}`"
       elsif version == previous_version &&
             package_manager == "docker"
         digest = docker_digest_from_reqs(requirements)
-        "`#{digest.split(':').last[0..6]}`"
+        "`#{T.must(digest.split(':').last)[0..6]}`"
       else
         version
       end
     end
 
+    sig { params(requirements: T::Array[T::Hash[Symbol, T.untyped]]).returns(String) }
     def docker_digest_from_reqs(requirements)
       requirements
         .filter_map { |r| r.dig(:source, "digest") || r.dig(:source, :digest) }
         .first
```

Comment:
Ooh, good catch. Let me double check this.

Reply:
Ah, it's because `requirements` values are `T.untyped`, which is an escape hatch. Ideally I'd define a specific object with specific properties, instead of use `T::Hash`.